src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,757 @@
+                                      const Ip6::Address &aSrcAddress,
+                                      const uint8_t *     aBuffer,
+                                      uint16_t            aBufferLength)
+{
+    const Ip6::Icmp::Header *icmp6Header;
+
+    // The Border Routing function is enabled only when we are attached.
+    VerifyOrExit(IsInitialized() && Get<Mle::MleRouter>().IsAttached());
+
+    VerifyOrExit(aInfraIfIndex == mInfraIfIndex);
+    VerifyOrExit(aBuffer != nullptr && aBufferLength >= sizeof(*icmp6Header));
+
+    icmp6Header = reinterpret_cast<const Ip6::Icmp::Header *>(aBuffer);
+
+    switch (icmp6Header->GetType())
+    {
+    case Ip6::Icmp::Header::kTypeRouterAdvert:
+        HandleRouterAdvertisement(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    case Ip6::Icmp::Header::kTypeRouterSolicit:
+        HandleRouterSolicit(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    default:
+        ExitNow();
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(IsInitialized());
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+    if (aEvents.Contains(kEventThreadNetdataChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            EvaluateRoutingPolicy();
+        }
+    }
+
+exit:
+    return;
+}
+
+uint8_t RoutingManager::EvaluateOmrPrefix(Ip6::Prefix *aNewOmrPrefixes, uint8_t aMaxOmrPrefixNum)
+{
+    uint8_t                         newOmrPrefixNum = 0;
+    NetworkData::Iterator           iterator        = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
+    bool                            isLocalOmrPrefixPublished = false;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        if (onMeshPrefixConfig.GetPrefix() == mLocalOmrPrefix)
+        {
+            // We didn't immediately add local OMR prefix here so that
+            // we don't need to remove it later when we want to unpublish it.
+            isLocalOmrPrefixPublished = true;
+        }
+        else if (newOmrPrefixNum < aMaxOmrPrefixNum)
+        {
+            aNewOmrPrefixes[newOmrPrefixNum++] = onMeshPrefixConfig.GetPrefix();
+        }
+        else
+        {
+            otLogWarnBr("EvaluateOmrPrefix: too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+        }
+    }
```

Comment:
In `EvaluateOmrPrefix()` we go through the network data and get all on-mesh prefix entries and then check the ones that match the OMR conditions (slaac, default-route, etc) and add them in a array. We later use the array when we prepare RA msg and include all the prefixes as "RouteInfoOption" (which is send on infra netif).

It's valid/possible that the same prefix is added by one more than one thread device (we may even want to do this for OMR). When we do `GetNextOnMeshPrefix()` we get an entry for every thread devcie that has added the prefix (the `config` will have same prefix but different `mRloc`  which indicates who added this prefix).

I think we should check for this situation. Otherwise we may end up with duplicate entries in the array (same prefix) and duplicate RouteInfo Options in the RA msg.

Reply:
Good catch, I think I missed the situation that multiple prefixes may be the same. Will check duplicated prefixes.

Reply:
done with [this](https://github.com/openthread/openthread/pull/5856/commits/331ab8fd2523ae49898985e4f0d95b89f301542f) commit.